DllImportGenerator/DllImportGenerator/Marshalling/StringMarshaller.PlatformDefined.cs

```diff
@@ -0,0 +1,174 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using static Microsoft.CodeAnalysis.CSharp.SyntaxFactory;
+
+using static Microsoft.Interop.MarshallerHelpers;
+
+namespace Microsoft.Interop
+{
+    internal sealed class PlatformDefinedStringMarshaller : ConditionalStackallocMarshallingGenerator
+    {
+        private static readonly TypeSyntax NativeType = PointerType(PredefinedType(Token(SyntaxKind.VoidKeyword)));
+
+        private readonly IMarshallingGenerator windowsMarshaller;
+        private readonly IMarshallingGenerator nonWindowsMarshaller;
+
+        public PlatformDefinedStringMarshaller(IMarshallingGenerator windowsMarshaller, IMarshallingGenerator nonWindowsMarshaller)
+        {
+            this.windowsMarshaller = windowsMarshaller;
+            this.nonWindowsMarshaller = nonWindowsMarshaller;
+        }
+
+        public override ArgumentSyntax AsArgument(TypePositionInfo info, StubCodeContext context)
+        {
+            return Argument(
+                ConditionalExpression(
+                    IsWindows,
+                    this.windowsMarshaller.AsArgument(info, context).Expression,
+                    this.nonWindowsMarshaller.AsArgument(info, context).Expression));
+        }
```

Comment:
I don't know if the Roslyn syntax APIs provide a way to do equality checks, but if possible it would be nice to not emit the conditional expression if both arguments would be the same. This would help readability and would reduce how much work the JIT has to do when it optimizes away the `OperatingSystem.IsWindows()` calls.

Reply:
There is `SyntaxNode.IsEquivalentTo` - I'll take a look at using that.

Reply:
Avoiding the conditional if the arguments are the same now (updated the generated code example in the description as well).

I only did it for the arguments here - not the conditional code blocks for marshalling / unmarshalling, since at this point those would never be the same, so doing an equivalence check for all the statements would be a bunch of extra unnecessary work.